File: src/main/java/official/kirizu/kirivers/client/Updater.java
package official.kirizu.kirivers.client;

import official.kirizu.kirivers.client.adapter.DeltaMagic;
import official.kirizu.kirivers.client.adapter.JdkSignatureVerifier;
import official.kirizu.kirivers.client.adapter.SignatureVerifier;
import official.kirizu.kirivers.client.model.ClientLoginInput;
import official.kirizu.kirivers.client.model.Diff200;
import official.kirizu.kirivers.client.model.DiffRequest;
import official.kirizu.kirivers.client.model.FileEntry;
import official.kirizu.kirivers.client.model.Integrity200;
import official.kirizu.kirivers.client.model.IntegrityQuery;
import official.kirizu.kirivers.client.model.Pack200;
import official.kirizu.kirivers.client.model.PackRequest;
import official.kirizu.kirivers.client.model.TelemetryRequest;
import official.kirizu.kirivers.client.model.UpdateCheck200;
import official.kirizu.kirivers.client.model.UpdateCheckRequest;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;

/**
 * Optional high-level Update: check → download/diff/pack → verify → optional patch/unpack/replace.
 * Telemetry failures never fail the result. Missing Replacer returns a staged path.
 */
public final class Updater {

  private final Client client;

  public Updater(Client client) {
    this.client = Objects.requireNonNull(client, "client");
  }

  public UpdateResult update(UpdateRequest request) throws IOException {
    Objects.requireNonNull(request, "request");
    if (request.currentVersion == null || request.os == null || request.arch == null) {
      throw new IllegalArgumentException("currentVersion, os and arch are required");
    }
    Config cfg = client.config();
    if (request.reportDevice && request.deviceId != null && !request.deviceId.isBlank()) {
      ClientLoginInput login = new ClientLoginInput();
      login.deviceId = request.deviceId;
      login.version = request.currentVersion;
      login.os = request.os;
      login.arch = request.arch;
      login.channel = request.channel;
      login.custom = request.custom;
      try {
        client.deviceReport(login);
      } catch (ApiException ignored) {
        // report is optional; check still proceeds
      }
    }

    UpdateCheckRequest checkReq = new UpdateCheckRequest();
    checkReq.currentVersion = request.currentVersion;
    checkReq.os = request.os;
    checkReq.arch = request.arch;
    checkReq.channel = request.channel;
    checkReq.hwRev = request.hwRev;
    checkReq.osVersion = request.osVersion;
    checkReq.deviceId = request.deviceId;
    CheckOutcome check = client.check(checkReq, request.ifNoneMatch);
    if (check.status() == CheckOutcome.Status.NOT_MODIFIED) {
      return new UpdateResult(UpdateResult.Status.NOT_MODIFIED, check, null, null, null, false);
    }
    if (check.status() == CheckOutcome.Status.NO_UPDATE || !check.hasUpdate()) {
      return new UpdateResult(UpdateResult.Status.UP_TO_DATE, check, null, null, null, false);
    }

    UpdateCheck200 body = check.body();
    String target = body.versionSemver != null ? body.versionSemver : String.valueOf(body.versionInteger);
    String diffMode = "full_package";
    Path staged;
    String sha;
    try {
      report(request, target, "downloading", diffMode, null, null);
      Downloaded downloaded = downloadAndVerify(request, body);
      diffMode = downloaded.diffMode;
      staged = downloaded.path;
      sha = downloaded.sha256;
      if (cfg.replacer() != null && request.targetPath != null) {
        report(request, target, "applying", diffMode, null, null);
        cfg.replacer().replace(staged, request.targetPath);
        report(request, target, "installed", diffMode, null, null);
        return new UpdateResult(UpdateResult.Status.REPLACED, check, request.targetPath, sha, diffMode, true);
      }
      report(request, target, "installed", diffMode, null, null);
      return new UpdateResult(UpdateResult.Status.STAGED, check, staged, sha, diffMode, false);
    } catch (Exception e) {
      report(request, target, "failed", diffMode, "UPDATE_FAILED", e.getMessage());
      if (e instanceof IOException io) {
        throw io;
      }
      if (e instanceof RuntimeException re) {
        throw re;
      }
      throw new IOException(e);
    }
  }

  private Downloaded downloadAndVerify(UpdateRequest request, UpdateCheck200 check) throws IOException {
    Config cfg = client.config();
    boolean downgrade = Boolean.TRUE.equals(check.isDowngrade);
    boolean multi = "multi_file".equalsIgnoreCase(check.packageType);

    if (!multi
        && !downgrade
        && Boolean.TRUE.equals(check.deltaAvailable)
        && cfg.patcher() != null
        && localSha(request) != null) {
      try {
        return applyDelta(request, check);
      } catch (Exception ignored) {
        // unknown magic / patch failure → full package
      }
    }

    if (multi && cfg.fileStore() != null && request.installDir != null) {
      try {
        return applyPack(request, check);
      } catch (Exception ignored) {
        // fall back to full zip
      }
    }

    return downloadFull(request, check, "full_package");
  }

  private Downloaded applyDelta(UpdateRequest request, UpdateCheck200 check) throws IOException {
    Config cfg = client.config();
    DiffRequest diffReq = new DiffRequest();
    diffReq.sourceVersion = request.currentVersion;
    diffReq.targetVersion = check.versionSemver != null ? check.versionSemver : String.valueOf(check.versionInteger);
    diffReq.os = request.os;
    diffReq.arch = request.arch;
    diffReq.channel = request.channel != null ? request.channel : check.targetChannel;
    diffReq.hwRev = request.hwRev;
    diffReq.deviceId = request.deviceId;
    diffReq.localSha256 = localSha(request);
    diffReq.capabilities = new ArrayList<>(cfg.capabilities());
    diffReq.acceptedDeltaAlgos = new ArrayList<>(cfg.acceptedDeltaAlgos());
    Diff200 diff = client.diff(diffReq);
    if (diff.diffMode == null || !"binary_delta".equals(diff.diffMode) || diff.packageUrl == null) {
      return downloadFull(request, check, "full_package");
    }
    byte[] delta = client.downloadUrl(diff.packageUrl, null).body();
    DeltaMagic.requireSupported(cfg.patcher().supportedAlgos(), delta);
    byte[] source = request.localFile;
    if (source == null) {
      throw new IOException("localFile is required to apply a binary delta");
    }
    byte[] patched = cfg.patcher().apply(source, delta);
    String expect = check.sha256;
    verifySha(patched, expect);
    verifySignature(
        check.versionInteger,
        check.versionSemver,
        check.rootHash,
        check.packageUrl,
        check.size,
        check.sha256,
        check.signature);
    Path dest = stageFile(request, check.fileName, patched);
    return new Downloaded(dest, expect, "binary_delta");
  }

  private Downloaded applyPack(UpdateRequest request, UpdateCheck200 check) throws IOException {
    Config cfg = client.config();
    String target = check.versionSemver != null ? check.versionSemver : String.valueOf(check.versionInteger);
    IntegrityQuery iq = new IntegrityQuery();
    iq.os = request.os;
    iq.arch = request.arch;
    iq.channel = request.channel != null ? request.channel : check.targetChannel;
    iq.hwRev = request.hwRev;
    iq.hashAlgo = "sha256";
    Integrity200 integrity = client.integrity(target, iq).body();
    if (integrity == null) {
      return downloadFull(request, check, "full_package");
    }
    List<String> needed = neededPaths(request.installDir, integrity.files);
    PackRequest packReq = new PackRequest();
    packReq.sourceVersion = request.currentVersion;
    packReq.targetVersion = target;
    packReq.os = request.os;
    packReq.arch = request.arch;
    packReq.channel = request.channel != null ? request.channel : check.targetChannel;
    packReq.hwRev = request.hwRev;
    packReq.deviceId = request.deviceId;
    packReq.neededPaths = needed;
    PackOutcome pack = client.packUntilReady(packReq);
    Pack200 body = pack.body();
    if (pack.fullPackage() || body == null || body.packageUrl == null) {
      return downloadFull(request, check, "full_package");
    }
    byte[] zip = client.downloadUrl(body.packageUrl, null).body();
    verifySha(zip, body.sha256);
    verifySignature(
        body.versionInteger != null ? body.versionInteger : check.versionInteger,
        body.versionSemver != null ? body.versionSemver : check.versionSemver,
        body.rootHash != null ? body.rootHash : check.rootHash,
        body.packageUrl,
        body.size,
        body.sha256,
        body.signature);
    Path destDir = request.stageDir;
    if (destDir == null) {
      throw new IOException("stageDir is required");
    }
    Files.createDirectories(destDir);
    if (cfg.archiveUnpacker() != null && body.files != null) {
      Map<String, String> map = new LinkedHashMap<>();
      for (FileEntry f : body.files) {
        if (f.sha256 != null && f.path != null) {
          map.put(f.sha256, PathUtil.normalize(f.path));
        }
      }
      cfg.archiveUnpacker().unpack(zip, map, destDir);
      return new Downloaded(destDir, body.sha256, "patch_package");
    }
    Path file = stageFile(request, body.fileName, zip);
    return new Downloaded(file, body.sha256, "patch_package");
  }

  private Downloaded downloadFull(UpdateRequest request, UpdateCheck200 check, String mode) throws IOException {
    if (check.packageUrl == null || check.packageUrl.isBlank()) {
      throw new IOException("check response missing package_url");
    }
    byte[] bytes = client.downloadUrl(check.packageUrl, null).body();
    verifySha(bytes, check.sha256);
    verifySignature(
        check.versionInteger,
        check.versionSemver,
        check.rootHash,
        check.packageUrl,
        check.size,
        check.sha256,
        check.signature);
    Path dest = stageFile(request, check.fileName, bytes);
    return new Downloaded(dest, check.sha256, mode);
  }

  private Path stageFile(UpdateRequest request, String fileName, byte[] bytes) throws IOException {
    Path dir = request.stageDir;
    if (dir == null) {
      throw new IOException("stageDir is required");
    }
    Files.createDirectories(dir);
    String name = (fileName == null || fileName.isBlank()) ? "package.bin" : PathUtil.normalize(fileName);
    if (name.contains("/")) {
      name = name.substring(name.lastIndexOf('/') + 1);
    }
    Path dest = dir.resolve(name);
    if (client.config().fileStore() != null) {
      client.config().fileStore().write(dest, bytes);
    } else {
      Files.write(dest, bytes);
    }
    return dest;
  }

  private void verifySha(byte[] bytes, String expected) throws IOException {
    if (expected == null || expected.isBlank() || client.config().hasher() == null) {
      return;
    }
    String got = client.config().hasher().sha256(bytes);
    if (!PathUtil.hashesEqual(got, expected)) {
      throw new IOException("sha256 mismatch");
    }
  }

  private void verifySignature(
      Long versionInteger,
      String versionSemver,
      String rootHash,
      String packageUrl,
      Long size,
      String sha,
      String signature) {
    Config cfg = client.config();
    if (signature == null || signature.isBlank() || cfg.signingKeys().isEmpty() || cfg.signatureVerifier() == null) {
      return;
    }
    byte[] payload =
        JdkSignatureVerifier.buildCheckPayload(
            JdkSignatureVerifier.decimalOrEmpty(versionInteger),
            versionSemver,
            rootHash,
            packageUrl,
            JdkSignatureVerifier.decimalOrEmpty(size),
            sha);
    SignatureVerifier.SignatureException last = null;
    for (var key : cfg.signingKeys()) {
      try {
        cfg.signatureVerifier().verify(key.algo(), key.publicKeyPem(), payload, signature);
        return;
      } catch (SignatureVerifier.SignatureException e) {
        last = e;
      }
    }
    throw new ApiException(0, "SIGNATURE_MISMATCH", last == null ? "signature mismatch" : last.getMessage(), null, null);
  }

  private List<String> neededPaths(Path installDir, List<FileEntry> files) throws IOException {
    Set<String> needed = new LinkedHashSet<>();
    if (files == null) {
      return List.of();
    }
    for (FileEntry f : files) {
      if (f.path == null) {
        continue;
      }
      String rel = PathUtil.normalize(f.path);
      Path local = installDir;
      for (String part : rel.split("/")) {
        if (!part.isEmpty()) {
          local = local.resolve(part);
        }
      }
      boolean keep = PathUtil.isKeepIfExists(f.installPolicy);
      boolean exists = client.config().fileStore() != null && client.config().fileStore().exists(local);
      if (keep && exists) {
        boolean checkHash = Boolean.TRUE.equals(f.integrityCheck);
        if (!checkHash) {
          continue;
        }
        if (f.sha256 != null && PathUtil.hashesEqual(client.config().fileStore().sha256(local), f.sha256)) {
          continue;
        }
      } else if (exists && f.sha256 != null && PathUtil.hashesEqual(client.config().fileStore().sha256(local), f.sha256)) {
        continue;
      }
      needed.add(rel);
    }
    return new ArrayList<>(needed);
  }

  private String localSha(UpdateRequest request) {
    if (request.localSha256 != null && !request.localSha256.isBlank()) {
      return request.localSha256;
    }
    if (request.localFile != null && client.config().hasher() != null) {
      return client.config().hasher().sha256(request.localFile);
    }
    return null;
  }

  private void report(
      UpdateRequest request, String toVersion, String status, String diffMode, String errorCode, String errorMessage) {
    String channel = request.channel == null || request.channel.isBlank() ? "stable" : request.channel;
    TelemetryRequest tel = new TelemetryRequest();
    tel.os = request.os;
    tel.arch = request.arch;
    tel.channel = channel;
    tel.fromVersion = request.currentVersion;
    tel.toVersion = toVersion;
    tel.status = status;
    tel.deviceId = request.deviceId;
    tel.diffMode = diffMode;
    tel.errorCode = errorCode;
    tel.errorMessage = errorMessage;
    try {
      client.reportTelemetry(tel);
    } catch (RuntimeException ignored) {
      // never block apply
    }
  }

  private static final class Downloaded {
    final Path path;
    final String sha256;
    final String diffMode;

    Downloaded(Path path, String sha256, String diffMode) {
      this.path = path;
      this.sha256 = sha256;
      this.diffMode = diffMode;
    }
  }
}

File: src/main/kotlin/official/kirizu/kirivers/client/Updater.kt
package official.kirizu.kirivers.client

import kotlinx.serialization.json.JsonObject
import java.nio.file.Path
import java.time.Duration

data class UpdateIdentity(
    val currentVersion: String,
    val os: String,
    val arch: String,
    val channel: String = "stable",
    val deviceId: String? = null,
    val hwRev: String? = null,
    val osVersion: String? = null,
    val custom: JsonObject? = null,
    val reportDevice: Boolean = false,
    val localPackage: ByteArray? = null,
    val installRoot: Path? = null,
)

enum class UpdateKind {
    UP_TO_DATE,
    STAGED,
    APPLIED,
}

data class UpdateOutcome(
    val kind: UpdateKind,
    val check: UpdateCheck? = null,
    val stagedPath: String? = null,
    val bytes: ByteArray? = null,
    val sha256: String? = null,
    val applied: Boolean = false,
    val fallbackToFull: Boolean = false,
    val diffMode: String? = null,
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is UpdateOutcome) return false
        return kind == other.kind &&
            check == other.check &&
            stagedPath == other.stagedPath &&
            bytes.contentEquals(other.bytes) &&
            sha256 == other.sha256 &&
            applied == other.applied &&
            fallbackToFull == other.fallbackToFull &&
            diffMode == other.diffMode
    }

    override fun hashCode(): Int {
        var result = kind.hashCode()
        result = 31 * result + (check?.hashCode() ?: 0)
        result = 31 * result + (stagedPath?.hashCode() ?: 0)
        result = 31 * result + (bytes?.contentHashCode() ?: 0)
        result = 31 * result + (sha256?.hashCode() ?: 0)
        result = 31 * result + applied.hashCode()
        result = 31 * result + fallbackToFull.hashCode()
        result = 31 * result + (diffMode?.hashCode() ?: 0)
        return result
    }
}

/**
 * High-level check → download → verify → optional patch/unpack → optional replace.
 * Missing [Replacer] is not a failure: verified bytes are returned as staged.
 * Telemetry errors never fail the outcome. Capabilities follow D13.
 */
class Updater(
    private val client: Client,
    private val fileStore: FileStore? = null,
    private val hasher: Hasher = JdkHasher(),
    private val archiveUnpacker: ArchiveUnpacker? = JdkZipUnpacker(),
    private val signatureVerifier: SignatureVerifier? = JdkSignatureVerifier(),
    private val patcher: Patcher? = null,
    private val replacer: Replacer? = null,
    private val signingKeys: List<SigningKey> = emptyList(),
    private val packOptions: PackPollOptions = PackPollOptions(),
    private val stageName: String = "package.bin",
) {
    fun capabilities(): Pair<List<String>, List<String>?> {
        val caps = linkedSetOf(Capability.FULL_PACKAGE)
        if (fileStore?.canWriteIndividualFiles() == true) caps += Capability.FILE_LIST
        if (archiveUnpacker != null) caps += Capability.PATCH_PACKAGE
        val algos = patcher?.supportedAlgos().orEmpty().filter { it.isNotBlank() }
        if (algos.isNotEmpty()) {
            caps += Capability.BINARY_DELTA
            return caps.toList() to algos
        }
        return caps.toList() to null
    }

    fun run(identity: UpdateIdentity): UpdateOutcome {
        if (identity.reportDevice && !identity.deviceId.isNullOrEmpty()) {
            runCatching {
                client.reportDevice(
                    DeviceReportRequest(
                        deviceId = identity.deviceId,
                        os = identity.os,
                        arch = identity.arch,
                        channel = identity.channel,
                        version = identity.currentVersion,
                        custom = identity.custom,
                    ),
                )
            }
        }
        val (caps, algos) = capabilities()
        val checked = client.check(
            CheckRequest(
                currentVersion = identity.currentVersion,
                os = identity.os,
                arch = identity.arch,
                channel = identity.channel,
                hwRev = identity.hwRev,
                osVersion = identity.osVersion,
                deviceId = identity.deviceId,
                capabilities = caps,
                acceptedDeltaAlgos = algos,
            ),
        )
        val update = when (checked) {
            is CheckResult.UpToDate, is CheckResult.NotModified ->
                return UpdateOutcome(kind = UpdateKind.UP_TO_DATE)
            is CheckResult.Update -> checked.update
        }
        return try {
            val (bytes, sha, mode, fallback) = obtainBytes(identity, update, caps, algos)
            if (mode == "full_package") {
                verifyHash(sha, update.sha256)
            }
            verifySignature(update)
            val staged = writeStage(bytes, mode)
            val applied = applyReplace(staged, identity.installRoot)
            report(identity, update, TelemetryStatus.INSTALLED, mode)
            UpdateOutcome(
                kind = if (applied) UpdateKind.APPLIED else UpdateKind.STAGED,
                check = update,
                stagedPath = staged,
                bytes = bytes,
                sha256 = sha,
                applied = applied,
                fallbackToFull = fallback,
                diffMode = mode,
            )
        } catch (ex: Exception) {
            report(identity, update, TelemetryStatus.FAILED, error = ex)
            throw ex
        }
    }

    private data class Obtained(
        val bytes: ByteArray,
        val sha256: String,
        val mode: String,
        val fallback: Boolean,
    )

    private fun obtainBytes(
        identity: UpdateIdentity,
        update: UpdateCheck,
        caps: List<String>,
        algos: List<String>?,
    ): Obtained {
        val wantDelta = patcher != null &&
            update.deltaAvailable &&
            !update.isDowngrade &&
            update.packageType == "single_file" &&
            identity.localPackage != null
        var deltaAttempted = false
        if (wantDelta) {
            deltaAttempted = true
            try {
                return applyDelta(identity, update, caps, algos)
            } catch (_: Exception) {
                // Unknown magic, hash miss, or diff failure → full package.
            }
        }
        if (update.packageType == "multi_file" && fileStore != null && archiveUnpacker != null) {
            val packed = runPack(identity, update)
            if (packed != null) return packed
        }
        val body = client.download(update.packageUrl).body
        val hex = hasher.sha256Hex(body)
        if (update.sha256.isNotEmpty() && !hex.equals(update.sha256, ignoreCase = true)) {
            throw HashMismatchException(update.sha256, hex)
        }
        return Obtained(body, hex, "full_package", fallback = deltaAttempted)
    }

    private fun applyDelta(
        identity: UpdateIdentity,
        update: UpdateCheck,
        caps: List<String>,
        algos: List<String>?,
    ): Obtained {
        val live = patcher ?: throw UnknownDeltaMagicException("none")
        val local = identity.localPackage ?: throw UnknownDeltaMagicException("none")
        val localHash = hasher.sha256Hex(local)
        val target = update.versionSemver ?: update.versionInteger?.toString()
            ?: throw ApiException(400, "INVALID_REQUEST", "check response missing target version")
        val diff = client.diff(
            DiffRequest(
                sourceVersion = identity.currentVersion,
                targetVersion = target,
                os = identity.os,
                arch = identity.arch,
                channel = identity.channel,
                deviceId = identity.deviceId,
                hwRev = identity.hwRev,
                localSha256 = localHash,
                capabilities = caps,
                acceptedDeltaAlgos = algos,
            ),
        )
        if (diff.diffMode != "binary_delta" || diff.packageUrl.isNullOrEmpty()) {
            val body = client.download(update.packageUrl).body
            return Obtained(body, hasher.sha256Hex(body), "full_package", fallback = true)
        }
        val algo = diff.deltaAlgo ?: update.deltaAlgo ?: live.supportedAlgos().first()
        val delta = client.download(diff.packageUrl).body
        DeltaMagic.require(algo, delta)
        val patched = live.apply(local, delta, algo)
        val hex = hasher.sha256Hex(patched)
        if (update.sha256.isNotEmpty() && !hex.equals(update.sha256, ignoreCase = true)) {
            throw HashMismatchException(update.sha256, hex)
        }
        return Obtained(patched, hex, "binary_delta", fallback = false)
    }

    private fun runPack(identity: UpdateIdentity, update: UpdateCheck): Obtained? {
        val store = fileStore ?: return null
        val target = update.versionSemver ?: update.versionInteger?.toString() ?: return null
        val integrity = when (
            val cached = client.integrity(
                IntegrityQuery(
                    version = target,
                    os = identity.os,
                    arch = identity.arch,
                    channel = identity.channel,
                    hwRev = identity.hwRev,
                ),
            )
        ) {
            is Cached.NotModified -> return null
            is Cached.Fresh -> cached.value
        }
        val needed = neededPaths(integrity, store)
        val pack = client.packUntilReady(
            PackRequest(
                sourceVersion = identity.currentVersion,
                targetVersion = target,
                os = identity.os,
                arch = identity.arch,
                channel = identity.channel,
                deviceId = identity.deviceId,
                hwRev = identity.hwRev,
                neededPaths = needed,
            ),
            packOptions,
        )
        return when (pack.status) {
            "full_package" -> {
                val url = pack.packageUrl ?: update.packageUrl
                val body = client.download(url).body
                Obtained(body, hasher.sha256Hex(body), "full_package", fallback = true)
            }
            "ready" -> {
                val url = pack.packageUrl ?: return null
                val archive = client.download(url).body
                if (!pack.sha256.isNullOrEmpty()) {
                    val hex = hasher.sha256Hex(archive)
                    if (!hex.equals(pack.sha256, ignoreCase = true)) {
                        throw HashMismatchException(pack.sha256, hex)
                    }
                }
                val unpacker = archiveUnpacker ?: return null
                unpacker.unpack(archive, pack.files ?: integrity.files, store)
                Obtained(archive, hasher.sha256Hex(archive), "patch_package", fallback = false)
            }
            else -> null
        }
    }

    internal fun neededPaths(manifest: IntegrityManifest, store: FileStore): List<String> {
        val needed = LinkedHashSet<String>()
        for (file in manifest.files) {
            val path = try {
                InstallPaths.normalize(file.path)
            } catch (_: InvalidPathException) {
                continue
            }
            val local = store.read(path)
            val keep = file.installPolicy == "KEEP_IF_EXISTS" && local != null
            if (keep) {
                if (!file.integrityCheck) continue
                val expected = file.sha256
                if (expected.isNullOrEmpty()) continue
                if (hasher.sha256Hex(local).equals(expected, ignoreCase = true)) continue
            }
            needed += path
        }
        return needed.toList()
    }

    private fun verifyHash(actual: String, expected: String?) {
        if (expected.isNullOrEmpty()) return
        if (!actual.equals(expected, ignoreCase = true)) {
            throw HashMismatchException(expected, actual)
        }
    }

    private fun verifySignature(update: UpdateCheck) {
        val signature = update.signature ?: return
        val verifier = signatureVerifier ?: return
        if (signingKeys.isEmpty()) return
        val payload = CheckPayload.build(
            versionInteger = update.versionInteger,
            versionSemver = update.versionSemver,
            rootHash = update.rootHash,
            packageUrl = update.packageUrl,
            size = update.size,
            sha256 = update.sha256,
        ).toByteArray(Charsets.UTF_8)
        val ok = signingKeys.any { key ->
            runCatching {
                verifier.verify(key.algorithm, key.publicKey, payload, signature)
            }.getOrDefault(false)
        }
        if (!ok) throw SignatureVerifyException("check signature mismatch")
    }

    private fun writeStage(bytes: ByteArray, mode: String): String? {
        val store = fileStore ?: return null
        if (mode == "patch_package") {
            // Members are already unpacked into the store; do not drop the zip in as package.bin.
            return store.resolvePath("")?.toString()
        }
        store.write(stageName, bytes)
        return store.resolvePath(stageName)?.toString() ?: stageName
    }

    private fun applyReplace(stagedPath: String?, installRoot: Path?): Boolean {
        val live = replacer ?: return false
        val staged = stagedPath ?: return false
        val root = installRoot ?: return false
        live.replace(Path.of(staged), root)
        return true
    }

    private fun report(
        identity: UpdateIdentity,
        update: UpdateCheck,
        status: String,
        diffMode: String? = null,
        error: Exception? = null,
    ) {
        val to = update.versionSemver ?: update.versionInteger?.toString() ?: identity.currentVersion
        runCatching {
            client.reportTelemetry(
                TelemetryRequest(
                    os = identity.os,
                    arch = identity.arch,
                    channel = identity.channel,
                    fromVersion = identity.currentVersion,
                    toVersion = to,
                    status = status,
                    deviceId = identity.deviceId,
                    diffMode = diffMode,
                    errorCode = (error as? ApiException)?.code,
                    errorMessage = error?.message?.take(200),
                ),
            )
        }
    }

    companion object {
        fun packBackoffDelay(step: Int, initial: Duration = Duration.ofSeconds(1), cap: Duration = Duration.ofSeconds(15)): Duration {
            var ms = initial.toMillis()
            repeat(step) { ms = minOf(ms * 2, cap.toMillis()) }
            return Duration.ofMillis(ms)
        }
    }
}
